### PR TITLE
Extend schema validation to validate routing rules are valid

### PR DESCRIPTION
**Pull request by jonnyshaw89**
### What is the context of this PR?
Extend the schema validation test to ensure that the routing rules are semantically correct and cover all scenarios. 

Fixes 601 and Fixes 602

### How to review 
Re-test issues.

### CODE REVIEW

File: app/data/0_star_wars.json [hunk 1]
```diff
@@ -127,6 +127,11 @@
                                 }
                             }
 
+                        },
+                        {
+                            "goto": {
+                                "id": "96682325-47ab-41e4-a56e-8315a19ffe2a"
+                            }
```
Comment (samiwel) [inline app/data/0_star_wars.json:134]: Do we need to specify a `goto` route if it's just going to the next block? Assuming the previous routing rules don't fire, doesn't it just route to the next block anyway?

File: tests/app/data/test_schema_validation.py [hunk 2]
```diff
@@ -41,15 +47,67 @@ def test_schema(self):
...
         except JSONDecodeError as e:
             return "JSON Parse Error! Could not parse [{}]. Error [{}]".format(file, e)
 
 
+    @staticmethod
+    def validate_schema_contains_valid_routing_rules(file, json_to_validate):
+        for block in SchemaHelper.get_blocks(json_to_validate):
+            if 'routing_rules' in block and len(block['routing_rules']) > 0:
+                for rule in block['routing_rules']:
+                    is_goto_rule = 'goto' in rule and 'when' in rule['goto'].keys() or 'id' in rule['goto'].keys()
```
Comment (samiwel) [inline tests/app/data/test_schema_validation.py]: Do we need this check? `'when' in rule['goto'].keys() or ` Isn't it sufficient just to check for existence of `'id'`?

Comment (jonnyshaw89) [reply tests/app/data/test_schema_validation.py]: agree, ill change this

File: app/data/0_star_wars.json (hunk 1 above)
Comment (jonnyshaw89) [reply app/data/0_star_wars.json:134]: we need to specify, the way the navigator works if routing_rules are defined then 1 of them needs to match or default

Comment (samiwel) [reply app/data/0_star_wars.json:134]: Okay, fair enough.

File: tests/app/data/test_schema_validation.py [hunk 3]
```diff
@@ -28,11 +29,16 @@ def test_schema(self):
         for file in files:
 
             json_file = open(os.path.join(settings.EQ_SCHEMA_DIRECTORY, file), encoding="utf8")
+            json_to_validate = json.load(json_file)
+
+            error = self.validate_json_against_schema(file, json_to_validate, schema)
+
+            error = self.validate_schema_contains_valid_routing_rules(file, json_to_validate)
```
Comment (collisdigital) [inline tests/app/data/test_schema_validation.py]: Each time error is set is overriding the last before the list is extended finally below (will only lost last error found if there were multiple); maybe extend errors immediately? Or alternatively only call later checks of previous check had no errors?

Comment (collisdigital) [reply tests/app/data/test_schema_validation.py]: Only **list** ...

Comment (jonnyshaw89) [reply tests/app/data/test_schema_validation.py]: A keen eye you have.

File: tests/app/data/test_schema_validation.py [hunk 4]
```diff
@@ -41,15 +44,67 @@ def test_schema(self):
...
+                            if rule['goto']['id'] == 'summary':
+                                continue
+
+                            SchemaHelper.get_block(json_to_validate, rule['goto']['id'])
+                        except StopIteration:
+                            return "Schema Error. File[{}] Routing rule routes to invalid block [{}]".format(file, rule['goto']['id'])
+
+
+    @staticmethod
+    def validate_schema_containing_routing_rules_has_default_if_nots_all_answers_routed(file, json_to_validate):
```
Comment (benjefferies) [inline tests/app/data/test_schema_validation.py]: validate_schema_containing_routing_rules_has_default_if_nots_all_answers_routed -> validate_schema_containing_routing_rules_has_default_if_not_all_answers_routed

File: tests/app/data/test_schema_validation.py [hunk 5]
```diff
@@ -41,15 +44,72 @@ def test_schema(self):
...
+    def validate_schema_contains_valid_routing_rules(file, json_to_validate):
+
+        errors = []
+
+        for block in SchemaHelper.get_blocks(json_to_validate):
+            if 'routing_rules' in block and len(block['routing_rules']) > 0:
+                for rule in block['routing_rules']:
+                    is_goto_rule = 'goto' in rule and 'id' in rule['goto'].keys()
+                    if is_goto_rule:
+                        try:
```
Comment (benjefferies) [inline tests/app/data/test_schema_validation.py]: This try except block seems a bit nasty and hard to understand. I think we should create a has_block method in the test file instead

File: tests/app/data/test_schema_validation.py [hunk 6]
```diff
@@ -41,15 +44,72 @@ def test_schema(self):
...
+
+                            SchemaHelper.get_block(json_to_validate, rule['goto']['id'])
+                        except StopIteration:
+                            errors.append("Schema Error. File[{}] Routing rule routes to invalid block [{}]".format(file, rule['goto']['id']))
+
+        return errors
+
+
+    @staticmethod
+    def validate_schema_containing_routing_rules_has_default_if_nots_all_answers_routed(file, json_to_validate):
```
Comment (benjefferies) [inline tests/app/data/test_schema_validation.py]: validate_schema_containing_routing_rules_has_default_if_nots_all_answers_routed -> validate_schema_containing_routing_rules_has_default_if_not_all_answers_routed

File: tests/app/data/test_schema_validation.py [hunk 7]
```diff
@@ -41,15 +44,72 @@ def test_schema(self):
...
+
+        for block in SchemaHelper.get_blocks(json_to_validate):
+
+            if 'sections' in block:
+                for section in block['sections']:
+                    if 'questions' in section:
+                        for question in section['questions']:
+                            if 'answers' in question:
+                                for answer in question['answers']:
+                                    if answer['type'] == 'Radio' or answer['type'] == 'Checkbox':
```
Comment (benjefferies) [inline tests/app/data/test_schema_validation.py]: I think we should refactor this to a method called get_radio_and_checkbox_answers so it is more readable in what it's doing

File: tests/app/data/test_schema_validation.py [hunk 8]
```diff
@@ -41,15 +44,73 @@ def test_schema(self):
...
+
+                                        if 'routing_rules' in block and len(block['routing_rules']) > 0:
+
+                                            options = []
```
Comment (benjefferies) [inline tests/app/data/test_schema_validation.py]: Could do some nice bit of python magic here

```
options = [option['value'] for option in answer['options']]
```

File: tests/app/data/test_schema_validation.py [hunk 9]
```diff
@@ -41,15 +44,73 @@ def test_schema(self):
...
+                                                if 'goto' in rule:
+
+                                                    if 'when' in rule['goto'].keys():
+                                                        when = rule['goto']['when']
+                                                        if 'id' in when and when['id'] == answer['id'] and when['value'] in options:
+                                                            options.remove(when['value'])
+                                                    else:
+                                                        options = []
+
+                                            if len(options) > 0 and len(options) != len(answer['options']):
```
Comment (benjefferies) [inline tests/app/data/test_schema_validation.py]: Might be nice to refactor to boolean variable no_routing_or_all_routes for clarity

**Review by benjefferies — CHANGES_REQUESTED** [review]

**Review by benjefferies — APPROVED** [review]

File: tests/app/data/test_schema_validation.py [hunk 10]
```diff
@@ -28,27 +29,89 @@ def test_schema(self):
...
+
+    @staticmethod
+    def validate_schema_containing_routing_rules_has_default_if_not_all_answers_routed(file, json_to_validate):
+
+        errors = []
+
+        for block in SchemaHelper.get_blocks(json_to_validate):
+
+            if 'sections' in block:
+                for section in block['sections']:
```
Comment (ajmaddaford) [inline tests/app/data/test_schema_validation.py]: Can we remove some of the nesting by doing something like this for each loop:
```
for section in (s for s in block['sections'] if 'sections' in block)
```

File: tests/app/data/test_schema_validation.py [hunk 11]
```diff
@@ -28,27 +29,89 @@ def test_schema(self):
...
+                        if rule['goto']['id'] == 'summary':
+                            continue
+
+                        if not cls.has_block(json_to_validate, rule['goto']['id']):
+                            errors.append("Schema Error. File[{}] Routing rule routes to invalid block [{}]".format(file, rule['goto']['id']))
+
+        return errors
+
+    @staticmethod
+    def validate_schema_containing_routing_rules_has_default_if_not_all_answers_routed(file, json_to_validate):
```
Comment (ajmaddaford) [inline tests/app/data/test_schema_validation.py]: I think `validate_routing_rules_has_default_if_not_all_answers_routed` says the same in a few less words

File: tests/app/data/test_schema_validation.py [hunk 12]
```diff
@@ -28,27 +29,89 @@ def test_schema(self):
...
+    @classmethod
+    def validate_schema_contains_valid_routing_rules(cls, file, json_to_validate):
+
+        errors = []
+
+        for block in SchemaHelper.get_blocks(json_to_validate):
+            if 'routing_rules' in block and len(block['routing_rules']) > 0:
+                for rule in block['routing_rules']:
+                    is_goto_rule = 'goto' in rule and 'id' in rule['goto'].keys()
+                    if is_goto_rule:
```
Comment (ajmaddaford) [inline tests/app/data/test_schema_validation.py]: Can this not just be `if 'goto' in rule and 'id' in rule['goto'].keys():`?

File: tests/app/data/test_schema_validation.py [hunk 13]
```diff
@@ -28,27 +29,89 @@ def test_schema(self):
...
+                        if rule['goto']['id'] == 'summary':
+                            continue
+
+                        if not cls.has_block(json_to_validate, rule['goto']['id']):
```
Comment (ajmaddaford) [inline tests/app/data/test_schema_validation.py]: I think this method would be clearer as `schema_has_block`

File: tests/app/data/test_schema_validation.py [hunk 14]
```diff
@@ -28,27 +29,89 @@ def test_schema(self):
...
+    def validate_schema_contains_valid_routing_rules(cls, file, json_to_validate):
+
+        errors = []
+
+        for block in SchemaHelper.get_blocks(json_to_validate):
+            if 'routing_rules' in block and len(block['routing_rules']) > 0:
+                for rule in block['routing_rules']:
+                    is_goto_rule = 'goto' in rule and 'id' in rule['goto'].keys()
+                    if is_goto_rule:
+                        if rule['goto']['id'] == 'summary':
```
Comment (ajmaddaford) [inline tests/app/data/test_schema_validation.py]: `rule['goto']['id']` is used three times, might be neater using a variable e.g. `block_id`

**Review by ajmaddaford — CHANGES_REQUESTED** [review]

**Review by ajmaddaford — APPROVED** [review]

**Pull request merged**